File: backend/app/api/global_users.py
from fastapi import APIRouter, Depends, Query, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from app.core.database import get_db
from app.models.user import User
from app.models.bot import Bot
from app.models.telegram_user import TelegramUser
from app.schemas.telegram_user import TelegramUserResponse
from app.utils.dependencies import get_current_user

router = APIRouter(prefix="/users", tags=["users-global"])


@router.get("", response_model=List[TelegramUserResponse])
async def get_all_users(
    bot_id: Optional[int] = Query(None, description="Фильтр по ID бота"),
    status_filter: Optional[str] = Query(None, description="Фильтр по статусу (active, blocked, left)"),
    source_filter: Optional[str] = Query(None, description="Фильтр по источнику"),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Получить всех пользователей всех ботов пользователя, с возможностью фильтрации"""
    # Получаем все боты пользователя
    user_bots = db.query(Bot).filter(Bot.owner_id == current_user.id).all()
    bot_ids = [bot.id for bot in user_bots]
    
    if not bot_ids:
        return []
    
    # Запрос пользователей
    query = db.query(TelegramUser).filter(TelegramUser.bot_id.in_(bot_ids))
    
    # Применяем фильтр по bot_id, если указан
    if bot_id is not None:
        if bot_id not in bot_ids:
            return []  # Бот не принадлежит пользователю
        query = query.filter(TelegramUser.bot_id == bot_id)
    
    # Применяем фильтр по статусу
    if status_filter:
        query = query.filter(TelegramUser.status == status_filter)
    
    # Применяем фильтр по источнику
    if source_filter:
        query = query.filter(TelegramUser.source == source_filter)
    
    users = query.order_by(TelegramUser.joined_at.desc()).offset(skip).limit(limit).all()
    
    # Добавляем информацию о боте к каждому пользователю
    result = []
    for user in users:
        bot = next((b for b in user_bots if b.id == user.bot_id), None)
        bot_name = None
        if bot:
            bot_name = bot.name or bot.username or f"Bot #{bot.id}"
        
        # Создаем ответ с информацией о боте
        user_response = TelegramUserResponse(
            id=user.id,
            bot_id=user.bot_id,
            bot_name=bot_name,
            telegram_user_id=user.telegram_user_id,
            username=user.username,
            first_name=user.first_name,
            last_name=user.last_name,
            source=user.source,
            status=user.status,
            joined_at=user.joined_at,
            last_activity=user.last_activity,
            tags=[{"id": tag.id, "name": tag.name, "color": tag.color} for tag in user.tags]
        )
        result.append(user_response)
    
    return result

File: backend-node/src/api/globalUsers.ts
import express, { Request, Response } from 'express';
import { authenticateToken } from '../middleware/auth';
import prisma from '../core/database';

const router = express.Router();

function csvEscape(value: unknown): string {
  if (value === null || value === undefined) return '';
  const s = String(value);
  return /[",\n\r]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
}

function makeExportFilename(prefix: string): string {
  const d = new Date();
  const yyyy = d.getFullYear();
  const mm = String(d.getMonth() + 1).padStart(2, '0');
  const dd = String(d.getDate()).padStart(2, '0');
  return `${prefix}_${yyyy}-${mm}-${dd}.csv`;
}

/**
 * GET /api/users
 * Получить всех пользователей всех ботов пользователя, с возможностью фильтрации
 */

/**
 * GET /api/users/export
 * Выгрузить пользователей в CSV (по фильтрам: bot_id/status_filter/source_filter)
 */
router.get('/export', authenticateToken, async (req: Request, res: Response) => {
  try {
    const ownerId = req.user!.userId;
    const botId = req.query.bot_id ? parseInt(req.query.bot_id as string, 10) : undefined;
    const statusFilter = req.query.status_filter as string | undefined;
    const sourceFilter = req.query.source_filter as string | undefined;

    // Получаем все боты пользователя
    const userBots = await prisma.bot.findMany({
      where: { ownerId },
    });
    const botIds = userBots.map((bot) => bot.id);

    // Запрос пользователей
    const where: any = {
      botId: { in: botIds },
    };

    // Применяем фильтр по bot_id, если указан
    if (botId !== undefined) {
      if (!botIds.includes(botId)) {
        // Бот не принадлежит пользователю → отдаём пустой CSV
        res.setHeader('Content-Type', 'text/csv; charset=utf-8');
        res.setHeader('Content-Disposition', `attachment; filename="${makeExportFilename('users')}"`);
        res.setHeader('Cache-Control', 'no-store');
        res.write('\uFEFF');
        res.write('id,bot_id,bot_name,telegram_user_id,username,first_name,last_name,source,status,joined_at,last_activity,tags\n');
        res.end();
        return;
      }
      where.botId = botId;
    }

    // Применяем фильтр по статусу
    if (statusFilter) {
      where.status = statusFilter;
    }

    // Применяем фильтр по источнику
    if (sourceFilter) {
      where.source = sourceFilter;
    }

    const maxRows = parseInt(process.env.EXPORT_MAX_ROWS || '200000', 10);
    const total = botIds.length > 0 ? await prisma.telegramUser.count({ where }) : 0;
    if (Number.isFinite(maxRows) && maxRows > 0 && total > maxRows) {
      return res.status(413).json({
        detail: `Too many rows to export (${total}). Narrow filters or increase EXPORT_MAX_ROWS.`,
      });
    }

    const botNameById = new Map<number, string>();
    for (const b of userBots) {
      botNameById.set(b.id, b.name || b.username || `Bot #${b.id}`);
    }

    res.setHeader('Content-Type', 'text/csv; charset=utf-8');
    res.setHeader('Content-Disposition', `attachment; filename="${makeExportFilename('users')}"`);
    res.setHeader('Cache-Control', 'no-store');
    res.write('\uFEFF');
    res.write('id,bot_id,bot_name,telegram_user_id,username,first_name,last_name,source,status,joined_at,last_activity,tags\n');

    if (total === 0) {
      res.end();
      return;
    }

    const take = 5000;
    let lastId = 0;
    while (true) {
      const batch = await prisma.telegramUser.findMany({
        where: { ...where, id: { gt: lastId } },
        include: {
          tags: {
            select: { name: true },
          },
        },
        orderBy: { id: 'asc' },
        take,
      });

      if (batch.length === 0) break;

      for (const u of batch) {
        const row = [
          u.id,
          u.botId,
          botNameById.get(u.botId) || '',
          u.telegramUserId.toString(),
          u.username || '',
          u.firstName || '',
          u.lastName || '',
          u.source || '',
          u.status,
          u.joinedAt?.toISOString?.() ? u.joinedAt.toISOString() : String(u.joinedAt),
          u.lastActivity?.toISOString?.() ? u.lastActivity.toISOString() : String(u.lastActivity),
          (u.tags || []).map((t) => t.name).join(';'),
        ]
          .map(csvEscape)
          .join(',');

        res.write(row + '\n');
      }

      lastId = batch[batch.length - 1].id;
      if (batch.length < take) break;
    }

    res.end();
  } catch (error) {
    console.error('Error exporting users:', error);
    if (!res.headersSent) {
      return res.status(500).json({ detail: 'Internal server error' });
    }
    try {
      res.end();
    } catch {
      // ignore
    }
  }
});

router.get('/', authenticateToken, async (req: Request, res: Response) => {
  try {
    const userId = req.user!.userId;
    const botId = req.query.bot_id ? parseInt(req.query.bot_id as string, 10) : undefined;
    const statusFilter = req.query.status_filter as string | undefined;
    const sourceFilter = req.query.source_filter as string | undefined;
    const skip = parseInt((req.query.skip as string) || '0', 10);
    const limit = Math.min(parseInt((req.query.limit as string) || '100', 10), 1000);
    const includeTotal =
      req.query.include_total === '1' ||
      req.query.include_total === 'true';

    // Получаем все боты пользователя
    const userBots = await prisma.bot.findMany({
      where: { ownerId: userId },
    });

    const botIds = userBots.map((bot) => bot.id);

    if (botIds.length === 0) {
      return res.json(includeTotal ? { items: [], total: 0, skip, limit, counts: { active: 0, blocked: 0, left: 0 } } : []);
    }

    // Базовый where для счётчиков (без statusFilter)
    const whereBase: any = {
      botId: { in: botIds },
    };

    // Применяем фильтр по bot_id, если указан
    if (botId !== undefined) {
      if (!botIds.includes(botId)) {
        return res.json(includeTotal ? { items: [], total: 0, skip, limit, counts: { active: 0, blocked: 0, left: 0 } } : []); // Бот не принадлежит пользователю
      }
      whereBase.botId = botId;
    }

    // Применяем фильтр по источнику
    if (sourceFilter) {
      whereBase.source = sourceFilter;
    }

    // where для списка (с учетом statusFilter)
    const whereList = statusFilter ? { ...whereBase, status: statusFilter } : whereBase;

    const [users, total, grouped] = await Promise.all([
      prisma.telegramUser.findMany({
      where: whereList,
      include: {
        tags: {
          select: {
            id: true,
            name: true,
            color: true,
          },
        },
      },
      orderBy: { joinedAt: 'desc' },
      skip,
      take: limit,
      }),
      includeTotal ? prisma.telegramUser.count({ where: whereList }) : Promise.resolve(0),
      includeTotal
        ? prisma.telegramUser.groupBy({
            by: ['status'],
            where: whereBase,
            _count: { _all: true },
          })
        : Promise.resolve([] as Array<{ status: string; _count: { _all: number } }>),
    ]);

    const counts = { active: 0, blocked: 0, left: 0 } as Record<string, number>;
    if (includeTotal) {
      for (const g of grouped as any[]) {
        const key = String(g.status || '');
        const c = (g._count?._all ?? 0) as number;
        if (key in counts) counts[key] += c;
      }
    }

    // Добавляем информацию о боте к каждому пользователю
    const result = users.map((user) => {
      const bot = userBots.find((b) => b.id === user.botId);
      const botName = bot
        ? bot.name || bot.username || `Bot #${bot.id}`
        : null;

      return {
        id: user.id,
        bot_id: user.botId,
        bot_name: botName,
        telegram_user_id: Number(user.telegramUserId),
        username: user.username,
        first_name: user.firstName,
        last_name: user.lastName,
        source: user.source,
        status: user.status,
        joined_at: user.joinedAt,
        last_activity: user.lastActivity,
        tags: user.tags,
      };
    });

    return res.json(includeTotal ? { items: result, total, skip, limit, counts } : result);
  } catch (error) {
    console.error('Error getting all users:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

export default router;
